triton/tail.go

```diff
@@ -0,0 +1,236 @@
+package triton
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/tinylib/msgp/msgp"
+	"io"
+	"sync"
+	"time"
+)
+
+type TailAt struct {
+	stream            string
+	streamInited      bool
+	bucket            string
+	client            string
+	at                time.Time
+	closed            bool
+	records           chan Record
+	errors            chan error
+	region            string
+	kinesisService    KinesisService
+	s3Service         S3Service
+	pollInterval      time.Duration
+	emptyPollInterval time.Duration
+	sync.Mutex
+}
+
+// NewTailAtParams are the parameters for invoking NewTailAt
+type NewTailAtParams struct {
+	S3Service         S3Service
+	KinesisService    KinesisService
+	StreamName        string
+	Bucket            string
+	Client            string
+	At                time.Time
+	PollInterval      time.Duration
+	EmptyPollInterval time.Duration
+}
+
+const (
+	// The amount of time to poll for new kinesis records if we encounter an empty stream
+	DefaultEmptyPollInterval = 10 * time.Second
+	DefaultPollInterval      = 1 * time.Millisecond
+)
+
+// NewTail returns a new tailing stream starting at "at"
+func NewTailAt(params *NewTailAtParams) (tail *TailAt) {
+	if params.EmptyPollInterval == 0 {
+		params.EmptyPollInterval = DefaultEmptyPollInterval
+	}
+
+	if params.PollInterval == 0 {
+		params.PollInterval = DefaultPollInterval
+	}
+
+	tail = &TailAt{
+		at:                params.At,
+		emptyPollInterval: params.EmptyPollInterval,
+		kinesisService:    params.KinesisService,
+		pollInterval:      params.PollInterval,
+		s3Service:         params.S3Service,
+		stream:            params.StreamName,
+		bucket:            params.Bucket,
+		client:            params.Client,
+		records:           make(chan Record),
+		errors:            make(chan error),
+	}
+	return
+}
+
+func (t *TailAt) Next() (record Record, err error) {
+	t.Lock()
+	defer t.Unlock()
+	go t.initStream()
```

Comment:
I guess I don't really know this is a problem, but it seems somewhat wasteful to fire off a go routine for every record. Especially since mostly this i a no-op.